README.md
# Short Distributed ID generator module

This module was inspired by [icicle](https://github.com/intenthq/icicle) and [snowflake](https://github.com/twitter/snowflake). The ideas is to be able to generate non-colliding, URL friendly, and relatively short IDs that could be used in any application that requires to create URIs for arbitrary resources.

Looking around for what is available, I failed to find anything that would be simple and easy to implement. As a result, this module was born.

The id is a 64bit unsigned integer with 42 bits used for current timestamp in milliseconds, 10 bits used for shard id, and final 12 bits are used for revolving sequence.

| timestamp_ms | shard_id | sequence |
|:---:|:---:|:---:|
| 42bit | 10bit | 12bit |

## short-duid-js

[![Quality Gate Status](https://sonarcloud.io/api/project_badges/measure?project=phpb-com_short-duid-js&metric=alert_status)](https://sonarcloud.io/summary/new_code?id=phpb-com_short-duid-js)
[![npm version](https://img.shields.io/npm/v/short-duid-js.svg?style=flat-square)](https://www.npmjs.com/package/short-duid-js)
[![npm downloads](https://img.shields.io/npm/dm/short-duid-js.svg?style=flat-square)](https://www.npmjs.com/package/short-duid-js)

### Changelog
- 1.1.6 Depndencies' versions bumps and tidyup of variables
- 1.1.5 package.json clean-up and corrections
- 1.1.4 General clean-up and updates (no new functionality)
- 1.1.3 Updated dependencies
- 1.1.1 No changes to the code, fixing CI pipeline
- 1.1.0 Major update of dependencies
- 1.0.2 Minor fixes 
- 1.0.0 Initial release

### Requirements
- nodejs 12.X
- nodejs 14.X
- nodejs 16.X

### Features
- Written in pure JavaScript, no native code
- Time and sequence based numeric unique ID generation
- Time and sequence based alphanumeric URL-safe unique ID generation
- Designed to be distributed among 1024 shards, no need to synchronize runtime or after setup
- Can generate 4096 unique IDs per millisecond per shard (theoretically)
- Can generate unique IDs for 139 years without overflow or collision
- Resilient to time drift or sequence overflow, does not delay ID generation
- Allows to set custom epoch, prolong unique ID generation and shorten the ID
- Simple to use

### Installation
`npm install short-duid-js --save`

### How to use
This module is very simple to use, first you will need to initialize it and then start using its instance.

#### API

##### short_duid.init(shard_id, salt, epoch_start)
Instantiates short-duid and sets parameters for the life of instance.

###### Returns
- Short Distributed ID module instance or `Javascript Object` type
    - Initializes new instance of the module with the given parameters

###### Parameters
- `shard_id` - ID of this instance of short-duid, should be unique and not shared with other instances in the cluster; from 0 to 1023. This parameter will be converted into signed 32 bit integer and masked to fit in 12 bits.
- `salt` - Salt that is used by hashid encoder/decoder, should be constant and shared across all nodes in the cluster. Do not change this parameter once used in production, or you will have collisions in the alphanumeric IDs. Good way to generate salt on Linux: `dd if=/dev/random bs=1 count=102400 2>/dev/null| sha256sum`
- `epoch_start` - Number of **milliseconds** since unix epoch (1970, Jan 1 00:00:00 GMT). This should be some date in the near past and should never be changed further into the future once in production. Example: 1433116800000; //Mon, 01 Jun 2015 00:00:00 GMT. This parameter will be converted to unsigned 64bit integer.

____
##### _instance_.getDUID(count)
Method to retrieve array of DUIDs in alphanumeric form. Length of the array is specified by `count` parameter.

###### Returns
- `Javascript array` object of variable length, depending on `count` parameter.
    - Example: `[ "XLz0E3MvkEL" ]`

###### Parameters
- `count` - Number of alphanumeric DUIDs to return, from 0 to 8192.

____
##### _instance_.getDUIDInt(count)
Essential same method as `_instance_.getDUID` but instead of hashid converted integer, will return unique ID in a numeric form as string.

###### Returns
- `Javascript array` object of variable length, depending on `count` parameter.
    - Example: `[ "12534941854212112" ]`

###### Parameters
- `count` - Number of numeric DUIDs to return, from 0 to 8192.

____
##### _instance_.getShardID()
Method to get currently set shard ID of ShortDUID `_instance_`

###### Returns
- `number` current shard ID of ShortDUID `_instance_`
    - Example: `0`

###### Parameters
- `N/A`

____
##### _instance_.getEpochStart()
Method to get currently set custom epoch starting time in milliseconds of ShortDUID `_instance_`

###### Returns
- `string` currently set custom epoch of ShortDUID `_instance_`, since it is unsigned 64bit integer, we return it as string.
    - Example: `"0"`

###### Parameters
- `N/A`

____
##### _instance_.getSalt()
Method to get current salt of ShortDUID `_instance_`. Salt is used to generate alphanumeric DUIDs and also in `hashidEncode`/`hashidDecode` methods.

###### Returns
- `string` currently set salt that is used in hashid conversion of ShortDUID `_instance_`
    - Example: `"this is my salt"`

###### Parameters
- `N/A`

____
##### _instance_.hashidEncode(number_array)
Method to hash(encode) array of unsigned 64bit integers (in `Javascript string` format).

###### Returns
- `string` hashid of array of unsigned 64bit integers
    - Example: `"3nMMYV0PvMl"`

###### Parameters
- `number_array` - Array of unsigned 64bit integers in javascript number or string (if does not fit in `Javascript 58bit integer` data type) form.

____
##### _instance_.hashidDecode(hashid_string)
Decode previously encoded array of numbers with hashid method.

###### Returns
- `Javascript array` array of unsigned 64bit integers in a string form
    - Example: `['1', '2', '3']`

###### Parameters
- `hashid_string` - Hashid in a string form. Example: `3nMMYV0PvMl`

____
#### Advanced API
This API is mainly used by unit tests and should not be required for normal usage of the module. Use it at your own risk.

##### _instance_.getCurrentTimeMs()
Method to get current time since unix epoch in milliseconds as seen by the module, not adjusted for custom epoch. This method can be useful in testing and also in capturing _reference_ time to ensure timer stability across restarts.

###### Returns
- `string` of numbers, current time since unix epoch in milliseconds as seen by the module.

###### Parameters
- `N/A`

____
##### _instance_.driftTime(milliseconds)
Method to help simulate `system_clock` drift, can accept positive or negative integers.

###### Returns
- `string` number of milliseconds to drift ShortDUID's internal clock

###### Parameters
- `milliseconds` (optional) number of milliseconds to drift system_clock by, can be a positive or negative integer.


____
#### Example #1
Simplest example to execute all of the major methods of the module.
```javascript
var duid = require('short-duid-js');
var duid_instance = duid.init(0, "my salt", 0);
console.log(duid_instance.getDUID(10));
console.log(duid_instance.getDUIDInt(10));
```

#### Example #2
More complete example that will create API server with help of koajs and reply to queries.

##### package.json
```json
{
  "name": "ShortDUIDAPIServer",
  "description": "ShortDUID Example API Service",
  "version": "0.0.1",
  "private": true,
  "dependencies": {
    "koa": "*",
    "koa-bodyparser": "*",
    "koa-response-time": "*",
    "koa-router": "*",
    "koa-json": "*",
    "pm2": "*",
    "short-duid-js": "*"
  }
}
```

##### index.js
```javascript
'use strict';
var pm2 = require('pm2');

var cpus = require('os').cpus().length;
var procs = Math.ceil(0.75 * cpus);

pm2.connect(function() {
  pm2.start({
    name: 'ShortDUID',
    script: 'api_server.js',
    args: '',
    post_update: ['npm install --save', ' echo Launching ShortDUID'],
    node_args: ['--harmony', '--use_strict'],
    exec_interpreter: 'node',
    next_gen_js: false,
    exec_mode: 'cluster',
    min_uptime: '3600s',
    max_restarts: 25,
    cron_restart: '',
    instances: procs,
    max_memory_restart: '1G',
    error_file: './api_errors.log',
    out_file: './api_info.log',
    pid_file: './short-duid-api.pid',
    merge_logs: true,
    log_date_format: 'YYYY-MM-DD HH:mm:ss Z',
    vizion: true,
    autorestart: true,
    port: 65000,
    env: {
      NODE_ENV: 'production',
      AWS_KEY: 'XXX',
      AWS_SECRET: 'XXX'
    }
  }, function(err, apps) {
    if (err) console.log("Error: ", err, "App:", apps);
    pm2.disconnect();
  });
});
```

##### api_server.js
```javascript
'use strict';
var koa = require('koa');
var router = require('koa-router')();
var app = module.exports = koa();
var duid = require('short-duid-js');

//Define app name and port for koa-cluster
var cpus = require('os').cpus().length;
app.name = "ShortDUID";
app.node_id = 0;
app.nid = process.env.NODE_APP_INSTANCE ? process.env.NODE_APP_INSTANCE : (process.pid % cpus); //nodejs instance ID
app.shard_id = app.node_id + app.nid;
app.port = 65000;
app.salt = "this is my super secret salt";
app.epoch_start = 1433116800 * 1000; //Mon, 01 Jun 2015 00:00:00 GMT

//Instantiate short-duid
var duid_instance = new duid.init(app.shard_id, app.salt, app.epoch_start);

//Setup routes
router
  .get('/', function*(next) {
    this.body = {
      name: 'ShortDUID API'
    };
  })
  .get('/nduid/:count?', function*(next) {
    this.body = yield(duid_instance.getDUIDInt((this.params.count ? this.params.count : 1)));
  })
  .get('/duid/:count?', function*(next) {
    this.body = yield(duid_instance.getDUID((this.params.count ? this.params.count : 1)));
  });

//Setup middleware
app
  .use(require('koa-json')())
  .use(require('koa-response-time')())
  .use(router.routes())
  .use(router.allowedMethods());

app.listen(app.port);
```

And then you should install application with `npm install --save` and start the server by `node index.js`. You can check the logs and also list the processes with `./node_modules/.bin/pm2 list`. Getting fresh id can be done by curl: `curl http://localhost:65000/duid/`.

#### More examples
For more examples please see  `examples` folder, which I plan to keep adding to. You are free to contribute more examples.

### Projects using ShortDUID
So far I know of none, if you are using it in your project and do not mind sharing this information, please drop me a note at <ian@phpb.com>, and I will add you to this list.

### Testing
`git clone https://github.com/phpb-com/short-duid-js.git && cd short-duid-js && npm install --save-dev` <br />
`npm test && npm run bench`
```
> short-duid-js@1.1.3 test
> ./node_modules/mocha/bin/mocha --harmony --reporter spec ./test/

  Short DUID
    #getEpochStart()
      ✔ should return set epoch start, for instance #1: 1433116800000
      ✔ should return set epoch start, for instance #2: 1433116800000
      ✔ instance #1 and instance #2 should return same epoch start: 1433116800000
      ✔ should reset custom epoch to zero if given one larger than real epoch
      ✔ should accept custom epoch that is even 1 millisecond in the past
    #getSalt()
      ✔ should return set salt, for instance #1: 39622feb2b3e7aa7208f50f45ec36fd513baadad6977b53295a3b28aeaed4a54
      ✔ should return set salt, for instance #2: 39622feb2b3e7aa7208f50f45ec36fd513baadad6977b53295a3b28aeaed4a54
      ✔ instance #1 and instance #2 should return same salt: 39622feb2b3e7aa7208f50f45ec36fd513baadad6977b53295a3b28aeaed4a54
    #getShardID()
      ✔ should overflow if shard id is set to integer that does not fit in 10 bits: 1024 --> 0
      ✔ should overflow if shard id is set to integer that does not fit in 10 bits: 1025 --> 1
      ✔ should return set shard id for id that fits within 10 bits: 1023 --> 1023
      ✔ should return set shard id for id that fits within 10 bits: 0 --> 0
      ✔ should return set shard id for instance #1: 123
      ✔ should return set shard id for instance #2: 12
      ✔ should return different shard ids for instance #1 and instance #2
    #getDUID()
      ✔ Asked for 1 DUIDs, correctly returns 1 DUIDs
      ✔ Asked for 0 DUIDs, correctly returns 0 DUIDs
      ✔ Asked for 8192 DUIDs, correctly returns 8192 DUIDs (164ms)
      ✔ Asked for 8193 DUIDs, correctly returns 1 DUIDs
      ✔ should have no duplicates in the returned arrays, 8192 IDs each, and combined. (267ms)
    #getDUIDInt()
      ✔ Asked for 1 Int DUIDs, correctly returns 1 Integer DUIDs
      ✔ Asked for 0 Int DUIDs, correctly returns 0 Integer DUIDs
      ✔ Asked for 8192 Int DUIDs, correctly returns 8192 Integer DUIDs (86ms)
      ✔ Asked for 8193 Int DUIDs, correctly returns 1 Integer DUIDs
      ✔ should have no duplicates in the returned arrays, 8192 IDs each, and combined. (95ms)
    DUID with drifting time
      ✔ should return same drift time as given as parameter
      ✔ should generate ID with -1713 millisecond drift into the past from now( 202203420920 ), 848102616960839681 should be numerically smaller than 848102624363786241
      ✔ should consistently generate unique IDs even when time is drifting backwards constantly (371ms)


  28 passing (1s)


> Executing task: npm run bench <


> short-duid-js@1.1.3 bench
> /usr/bin/env node benchmarks/test.js

single DUIDInt generation x 288,177 ops/sec ±4.59% (77 runs sampled)
batch of 10 DUIDInt generation x 19,600 ops/sec ±11.33% (50 runs sampled)
single DUID generation x 70,656 ops/sec ±11.10% (61 runs sampled)
batch of 10 DUID generation x 10,967 ops/sec ±2.45% (86 runs sampled)
single DUID generation (1 character salt) x 108,127 ops/sec ±4.77% (86 runs sampled)
batch of 10 DUID generation (1 character salt) x 8,215 ops/sec ±23.60% (72 runs sampled)

```
## TODO
- [ ] Improve error handling, at the moment most of them are silent and prefer overflow
- [ ] Add more examples

## Contributing
All are welcome to submit issues, ideas, pull requests and patches

## License

The MIT License (MIT)

Copyright (c) 2015 Ian Matyssik <ian@phpb.com>

Permission is hereby granted, free of charge, to any person obtaining a copy of
this software and associated documentation files (the "Software"), to deal in
the Software without restriction, including without limitation the rights to
use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
the Software, and to permit persons to whom the Software is furnished to do so,
subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


CHANGELOG
- 1.1.6 Depndencies' versions bumps and tidyup of variables
- 1.1.5 package.json clean-up and corrections
- 1.1.4 General clean-up and updates (no new functionality)
- 1.1.3 Updated dependencies
- 1.1.1 No changes to the code, fixing CI pipeline
- 1.1.0 Major update of dependencies
- 1.0.2 Minor fixes 
- 1.0.0 Initial release


test/test.js
var duid = require( '../index' );
var test = require( "unit.js" );
var async = require( "async" );
var BN = require( "bn.js" );
var _ = require('lodash');

var check_duplicates = function ( arr ) {
  arr.sort();
  for ( var i = 1; i < arr.length; i++ ) {
    if ( arr[ i - 1 ] === arr[ i ] )
      return false;
  }
  return true;
};

describe( 'Short DUID', function () {

  var init = duid.init;
  var epoch_start = 1433116800 * 1000; //Mon, 01 Jun 2015 00:00:00 GMT
  var salt = '39622feb2b3e7aa7208f50f45ec36fd513baadad6977b53295a3b28aeaed4a54'; //dd if=/dev/random bs=1 count=102400 2>/dev/null| sha256sum

  var duid_instance1 = new init( 123, salt, epoch_start );
  var duid_instance2 = new init( 12, salt, epoch_start );

  describe( '#getEpochStart()', function () {

    it( 'should return set epoch start, for instance #1: ' + epoch_start, function () {
      test.value( duid_instance1.getEpochStart() )
          .isEqualTo( epoch_start );
    } );

    it( 'should return set epoch start, for instance #2: ' + epoch_start, function () {
      test.value( duid_instance2.getEpochStart() )
          .isEqualTo( epoch_start );
    } );

    it( 'instance #1 and instance #2 should return same epoch start: ' + epoch_start, function () {
      test.value( duid_instance1.getEpochStart() ).isIdenticalTo( duid_instance2.getEpochStart() );
    } );

    it( 'should reset custom epoch to zero if given one larger than real epoch', function () {
      var custom_epoch_overflow = (new Date()).getTime() + _.random(2000, 1000 * 1000);
      var duid_instance_overflow = new init( 0, salt, custom_epoch_overflow );
      test.value( duid_instance_overflow.getEpochStart() ).isEqualTo( 0 );
      test.value( duid_instance_overflow.getEpochStart() ).isNotEqualTo( custom_epoch_overflow );
    } );

    it( 'should accept custom epoch that is even 1 millisecond in the past', function () {
      var custom_epoch_near = (new Date()).getTime() - _.random(1, 2);
      var duid_instance_near = new init( 0, salt, custom_epoch_near );
      test.value( duid_instance_near.getEpochStart() ).isEqualTo( custom_epoch_near );
    } );

  } );

  describe( '#getSalt()', function () {

    it( 'should return set salt, for instance #1: ' + salt, function () {
      test.string( duid_instance1.getSalt() ).isIdenticalTo( salt );
    } );

    it( 'should return set salt, for instance #2: ' + salt, function () {
      test.string( duid_instance2.getSalt() ).isIdenticalTo( salt );
    } );

    it( 'instance #1 and instance #2 should return same salt: ' + salt, function () {
      test.string( duid_instance1.getSalt() ).isIdenticalTo( duid_instance2.getSalt() );
    } );

  } );

  describe( '#getShardID()', function () {

    var duid_instance_shardid_overflow = new init( 1024, salt, 0 );
    var duid_instance_shardid_overflow_plus_one = new init( 1025, salt, 0 );
    var duid_instance_shardid_tip = new init( 1023, salt, 0 );
    var duid_instance_shardid_bottom = new init( 0, salt, 0 );

    it( 'should overflow if shard id is set to integer that does not fit in 10 bits: 1024 --> 0', function () {
      test.number( duid_instance_shardid_overflow.getShardID() ).is( 0 );
    } );

    it( 'should overflow if shard id is set to integer that does not fit in 10 bits: 1025 --> 1', function () {
      test.number( duid_instance_shardid_overflow_plus_one.getShardID() ).is( 1 );
    } );

    it( 'should return set shard id for id that fits within 10 bits: 1023 --> 1023', function () {
      test.number( duid_instance_shardid_tip.getShardID() ).is( 1023 );
    } );

    it( 'should return set shard id for id that fits within 10 bits: 0 --> 0', function () {
      test.number( duid_instance_shardid_bottom.getShardID() ).is( 0 );
    } );

    it( 'should return set shard id for instance #1: 123', function () {
      test.number( duid_instance1.getShardID() ).is( 123 );
    } );

    it( 'should return set shard id for instance #2: 12', function () {
      test.number( duid_instance2.getShardID() ).is( 12 );
    } );

    it( 'should return different shard ids for instance #1 and instance #2', function () {
      test.number( duid_instance2.getShardID() ).isNot( duid_instance1.getShardID() );
    } );

  } );

  describe( '#getDUID()', function () {
    this.timeout(5000); //hashids are not particularly fast and short-duid in js is not speedy as well
    var tests = [ {
      args: 1,
      expected: 1
    }, {
      args: 0,
      expected: 0
    }, {
      args: 8192,
      expected: 8192
    }, {
      args: 8193,
      expected: 1
    } ];

    tests.forEach( function ( iter ) {
      it( 'Asked for ' + iter.args + ' DUIDs, correctly returns ' + iter.expected + ' DUIDs', function () {
        test.array( duid_instance1.getDUID( iter.args ) ).hasLength( iter.expected );
      } );
    } );

    it( 'should have no duplicates in the returned arrays, 8192 IDs each, and combined.', function () {
      var res1 = duid_instance1.getDUID( 8192 );
      var res2 = duid_instance2.getDUID( 8192 );
      var res3 = duid_instance1.getDUID( 8192 );
      test.array( _.uniq( res1 ) ).is( res1 );
      test.array( _.uniq( res2 ) ).is( res2 );
      test.array( _.uniq( res3 ) ).is( res3 );
      test.array( _.uniq( _.flattenDeep( [ res1, res2, res3 ] ) ) ).is( _.flattenDeep( [ res1, res2, res3 ] ) );
    } );

  } );

  describe( '#getDUIDInt()', function () {
    var tests = [ {
      args: 1,
      expected: 1
    }, {
      args: 0,
      expected: 0
    }, {
      args: 8192,
      expected: 8192
    }, {
      args: 8193,
      expected: 1
    } ];

    tests.forEach( function ( iter ) {
      it( 'Asked for ' + iter.args + ' Int DUIDs, correctly returns ' + iter.expected + ' Integer DUIDs', function () {
        test.array( duid_instance1.getDUID( iter.args ) ).hasLength( iter.expected );
      } );
    } );

    it( 'should have no duplicates in the returned arrays, 8192 IDs each, and combined.', function () {
      var res1 = duid_instance1.getDUIDInt( 8192 );
      var res2 = duid_instance2.getDUIDInt( 8192 );
      var res3 = duid_instance1.getDUIDInt( 8192 );
      test.array( _.uniq( res1 ) ).is( res1 );
      test.array( _.uniq( res2 ) ).is( res2 );
      test.array( _.uniq( res3 ) ).is( res3 );
      test.array( _.uniq( _.flattenDeep( [ res1, res2, res3 ] ) ) ).is( _.flattenDeep( [ res1, res2, res3 ] ) );
    } );

  } );

  describe( 'DUID with drifting time', function () {

    var duid_instance3 = new init( 123, salt, epoch_start );
    var id1 = new BN( duid_instance3.getDUIDInt( 1 )[ 0 ], 10 );
    var drift = duid_instance3.driftTime( Math.random() * 1000 * 10 * -1 | 0 );
    var id2 = new BN( duid_instance3.getDUIDInt( 4096 )[ 4095 ], 10 ); //Need to rollover sequence
    var curr_ms_time = new BN( duid_instance3.getCurrentTimeMs(), 10 );

    it( 'should return same drift time as given as parameter', function () {
      duid_instance3.driftTime( -123 )
      test.number( duid_instance3.driftTime() ).isEqualTo( -123 );
      duid_instance3.driftTime( 123 )
      test.number( duid_instance3.driftTime() ).isEqualTo( 123 );
      duid_instance3.driftTime( 0 )
      test.number( duid_instance3.driftTime() ).isEqualTo( 0 );
    } );

    it( 'should generate ID with ' + drift + ' millisecond drift into the past from now( ' + curr_ms_time + ' ), ' + id1 + ' should be numerically smaller than ' + id2, function () {
      test.bool( id2.cmp( id1 ) === 1 ).isTrue();
    } );

    it( 'should consistently generate unique IDs even when time is drifting backwards constantly', function () {
      drift = duid_instance3.driftTime( 0 ); //Reset drift back to 0
      var duids = [];
      for(var i = 0; i < 10; ++i) {
        duid_instance3.driftTime( drift );
        var start = (new Date()).getTime();
        duids[i] = duid_instance3.getDUIDInt( 8192 );
        drift = start - (new Date()).getTime();
        test.number( drift ).isLessThan( 0 );
        test.array( duids[i] ).hasLength( 8192 );
        test.array( _.uniq( duids[i] ) ).is( duids[i] );
      }
      var merged_uniq = _.uniq( _.flattenDeep( duids ) );
      test.array( merged_uniq ).is( _.flattenDeep( duids ) );
      test.array( merged_uniq ).hasLength( 8192 * 10 );
    } );

  } );

} );

// vim: syntax=cpp11:ts=2:sw=2


benchmarks/test.js
var Benchmark = require('benchmark'),
    ShortDUID = require('../.');

var suit = new Benchmark.Suite;

var duid = ShortDUID.init(0, "b130389689f522fa8b6664eb291083551ff0c00a4cf5a4905fdee8cd9063e55a", 1433116800000);
var duid_small_salt = ShortDUID.init(0, "a", 1433116800000);

suit.add('single DUIDInt generation', function () {
    duid.getDUIDInt(1);
});

suit.add('batch of 10 DUIDInt generation', function () {
    duid.getDUIDInt(10);
});

suit.add('single DUID generation', function () {
    duid.getDUID(1);
});

suit.add('batch of 10 DUID generation', function () {
    duid.getDUID(10);
});

suit.add('single DUID generation (1 character salt)', function () {
    duid_small_salt.getDUID(1);
});

suit.add('batch of 10 DUID generation (1 character salt)', function () {
    duid_small_salt.getDUID(10);
});

suit.on('cycle', function(event) {
      console.log(String(event.target));
})

// run async
suit.run({ 'async': true });


index.js
'use strict';
const BN = require('bn.js'),
    Hashids = require('hashids');

exports.init = ShortDUID;

function ShortDUID(shard_id, salt, epoch_start) {
    if (!(this instanceof ShortDUID)) return new ShortDUID(shard_id, salt, epoch_start);
    const duid = this;

    duid.shard_id = new BN(shard_id, 10).maskn(10);
    duid.salt = salt;

    const now = new BN(+new Date(), 10);
    duid.epoch_start = new BN(epoch_start, 10);
    if(now.cmp( duid.epoch_start ) === -1) {
        duid.epoch_start = new BN(0, 10);
    }

    duid.time_drift = new BN(0, 10);
    duid.hashids = new Hashids(duid.salt);

    duid.sequence = new BN(0, 10);
    duid.ts_sequence = [];

    return duid;
}

ShortDUID.prototype.getDUID = function (count) {
    let ret = [];
    let cnt;

    if ( count === 0 ) return [];

    if ( count > 8192 || count < 1 ) {
        cnt = 1;
    } else {
        cnt = count;
    }

    for(let i = 0; i < cnt; i++) {
        ret.push(this.hashids.encodeHex(this.getID().toString(16)));
    }

    return ret;
}

ShortDUID.prototype.getDUIDInt = function (count) {
    let ret = [];
    let cnt;

    if ( count === 0 ) return [];

    if ( count > 8192 || count < 0 ) {
        cnt = 1;
    } else {
        cnt = count;
    }

    for(let i = 0; i < cnt; i++) {
        ret.push(this.getID().toString(10));
    }

    return ret;
}

ShortDUID.prototype.getID = function () {
    const duid = this;

    // Calculate time part
    let now = new BN(+new Date(), 10).sub(duid.epoch_start.add(duid.time_drift)).maskn(42);

    // Calculate shard id part
    const shid = duid.shard_id.ushln(12).clone(); // Shift left by 12 bit

    // Calculate sequence part
    const seq = duid.sequence.iadd(new BN(1, 10)).maskn(12).clone();
    if(duid.ts_sequence[seq] !== undefined &&
        (duid.ts_sequence[seq].cmp(now) === 0 ||
         duid.ts_sequence[seq].cmp(now) === 1)
      ) { // Overflowing or drifting backwards
        now = duid.ts_sequence[seq].clone();
        now.iadd(new BN(1, 10));
        now.imaskn(42);
    }
    duid.ts_sequence[seq] = now.clone();
    now.iushln(22);

    // Calculate final ID
    return now.or(shid.or(seq));
}

ShortDUID.prototype.getShardID = function() {
    return parseInt(this.shard_id.toString(10), 10);
}

ShortDUID.prototype.getEpochStart = function() {
    return this.epoch_start.toString(10);
}

ShortDUID.prototype.getSalt = function() {
    return this.salt;
}

ShortDUID.prototype.getCurrentTimeMs = function() {
    const duid = this;
    const estart = duid.epoch_start.clone();
    let now = new BN(+new Date(), 10); // Current system time in milliseconds converted to BN object

    now.isub(estart).imaskn(42); // Calculate custom epoch and add/subtract drift time
    return now.toString(10);
}

ShortDUID.prototype.driftTime = function(drift) {
    if(drift !== undefined) {
        this.time_drift = new BN(drift, 10);
    }

    return parseInt(this.time_drift.toString(10), 10);
}


examples/API_Server/api_server.js
'use strict';
var koa = module.exports = require( 'koa' );
var router = require( 'koa-router' )();
var app = module.exports = koa();
var duid = require( 'short-duid-js' );
var cluster = require( 'cluster' );

//Define app name and port for koa-cluster
var cpus = require( 'os' ).cpus().length;
app.name = "ShortDUID";
app.node_id = 0;
app.nid = process.env.NODE_APP_INSTANCE ? process.env.NODE_APP_INSTANCE : ( cluster.worker.id ? cluster.worker.id : ( process.pid % cpus ) ); //nodejs instance ID
app.shard_id = app.node_id + app.nid;
app.port = 45000;
app.salt = "this is my super secret salt";
app.epoch_start = 1433116800 * 1000; //Mon, 01 Jun 2015 00:00:00 GMT

//Instantiate short-duid
var duid_instance = new duid.init( app.shard_id, app.salt, app.epoch_start );
console.log( "Node with shard_id #" + app.shard_id + " started." );

//Setup routes
router
  .get( '/', function* ( next ) {
    this.body = {
      name: 'ShortDUID API'
    };
  } )
  .get( '/nduid/:count?', function* ( next ) {
    this.body = yield( duid_instance.getDUIDInt( ( this.params.count ? this.params.count : 1 ) ) );
  } )
  .get( '/duid/:count?', function* ( next ) {
    this.body = yield( duid_instance.getDUID( ( this.params.count ? this.params.count : 1 ) ) );
  } );

//Setup middleware
app
  .use( require( 'koa-json' )() )
  .use( require( 'koa-response-time' )() )
  .use( router.routes() )
  .use( router.allowedMethods() );
